### ISSUE TITLE
[utils] Use template literal types to parse esquery selectors to automatically type complex selectors

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#4065]
Template literal types (https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) were released in TS 4.1. since their announcement people have done all sorts of crazy and wonderful things with them - from simple cases like auto prefixing keys to ridiculous things like SQL query parsers.

We should be able to leverage them as part of our ESLint types to automatically type the `node` argument of an ESLint rule selector.

We don't need to handle every crazy case (selectors can get pretty complicated!) but at least handling things like property existence checks, sibling checks, and the `:exit` special selector would be great.

For reference here are the docs on, and implementation of esquery selectors
https://github.com/estools/esquery

It's worth mentioning that if we find that template literal types are too slow or complex, an alternative might be to consider using codegen to generate a number of common cases statically. But I'd love to achieve this using template literal types if possible!

**Comment by JoshuaKGoldberg** [comment]
**YES**. This would be such a good type system challenge 😄 

This feels like something that would be broadly useful outside of typescript-eslint for writing vanilla ESLint rules in TypeScript... I don't (yet?) know how that would be doable, but something to keep in mind maybe.

> consider using codegen to generate a number of common cases statically

Perhaps an ESLint rule that parses the type from the selector and enforces the right type on the node?

**Comment by bradzacher** [comment]
> Perhaps an ESLint rule that parses the type from the selector and enforces the right type on the node?

That was what I was something I was thinking we could do as well, yeah!

That we we can handle some simpler cases with template literal types and lean on probably correct explicit types enforced via a lint rule.

We could also leverage codegen to generate more selector strings. Eg we could take the type defs and generate extra defs for every node like:
```ts
'VariableDeclarator[init]'?: (
  node: VariableDeclarator & {
    init: NonNullable<VariableDeclarator['init']>,
  },
)
```

**Comment by auvred** [comment]
I created `magic-esquery` as a side pet project (I was curious about implementing some type-level parser :smile:). It's a type-level ESQuery parser and matcher. So it basically allows to infer types of nodes that can be queries by some selector
https://github.com/auvred/magic-esquery

Perhaps this could be useful for typescript-eslint!

It properly infers types for all the selectors used in `@typescript-eslint/eslint-plugin`, `@stylistic/eslint-plugin` and `eslint-plugin-jest` rules!

See tests:
`@typescript-eslint/eslint-plugin` - [tests](https://github.com/auvred/magic-esquery/blob/0b69a302bc891f1d0392fec3550a58ed69b10d01/tests/ecosystem/tests/ts-eslint.matcher.generated-test.ts)
`@stylistic/eslint-plugin` - [tests](https://github.com/auvred/magic-esquery/blob/0b69a302bc891f1d0392fec3550a58ed69b10d01/tests/ecosystem/tests/eslint-stylistic.matcher.generated-test.ts)
`eslint-plugin-jest` - [test](https://github.com/auvred/magic-esquery/blob/0b69a302bc891f1d0392fec3550a58ed69b10d01/tests/ecosystem/tests/jest.matcher.generated-test.ts)

_[You can try it online if you want!](https://www.typescriptlang.org/play?target=99&moduleResolution=99&module=199&jsx=0&inlineSourceMap=true#code/JYWwDg9gTgLgBDAnmApnA3nAigVxVROAXzgDMoIQ4ByEAQwHNgBjAWhQGcBHPA6gKFCRYCZGkwAVAMoBRKRKgo0JcpRoABJKg7MowMDHYcANsAB2MAPRbOu-Yc4xFKanH78bcRRwCMcALzYvIgAPNQAsiggAEb4MgAeYN4cwBBm1AA0cNJyCkoAdAByEAAmKAB8-JaWcLUAegD87p7eAEwBQfihEVGxUAlJnClpANrMlGA4MCgl-k54ALqZ2bLyzkWlFVU19U0eYl6cAMwduF1hAFz0MMwAFpwAFJExcYnJqWYXZhAwD2MTUxmcygiwAlFkAMJ0YzGAbvNJwcpwfLMaHGJQjMx0EAoBag5Y5NYFYplSrVWpwRpAA)_

---

I hope this will be useful for ts-eslint. Perhaps we can use this project as a reference to implement the same functionality as built-in in ts-eslint's utils.. Or maybe we can use it as is. If you want, I can file a draft pr with the `magic-esquery` used in RuleCreator.

So if you have any thoughts about this, please let me know!

**Comment by bradzacher** [comment]
@auvred do we know what the typecheck perf hit is for the types?
I'm certainly very interested in having this as it greatly improves DevX and removes the need for many manual and possibly incorrect annotations.

**Comment by auvred** [comment]
> do we know what the typecheck perf hit is for the types?

I removed explicit type annotations for >90% of existing selectors in the `packages/eslint-plugin/rules/*`

Tested with `hyperfine` with 3 runs. On my machine the results are as follows:

|command|Baseline|With `magic-esquery`|
|---|---|---|
|`yarn typecheck`|`9.099 s ±  0.207 s`|`10.919 s ±  0.051 s`|
|`yarn lint`|`54.137 s ±  0.335 s`|`57.145 s ±  1.121 s`|

So it adds +2s to the `typecheck` and +3s to the `lint`

**Comment by bradzacher** [comment]
🤔 that's really not that bad overall for a marked improvement in devx.
roughly how many selectors was that 90%? I'm assuming it's a few hundred.

I think this is a good thing to be added as part of our standard offering!

**Comment by JoshuaKGoldberg** [comment]
cc @bmish from https://github.com/eslint-community/eslint-plugin-eslint-plugin/issues/339, thinking on what a rule to ban overly complex ESQuery selectors should consider "complex". I think it'd make sense to have _"whatever `magic-esquery` can support"_ be an option for what's considered reasonable.

**Comment by JoshuaKGoldberg** [comment]
cc @bmish from https://github.com/eslint-community/eslint-plugin-eslint-plugin/issues/339, thinking on what a rule to ban overly complex ESQuery selectors should consider "complex". I think it'd make sense to have _"whatever `magic-esquery` can support"_ be an option for what's considered reasonable.

Aside: `magic-esquery` is a very cute name (I find it very charming 😄) but not super descriptive. It'd be nice to have one that's more clear IMO. Especially since folks will inevitably want to dive into it to investigate type edge cases... `typed-esquery`? `esquery-types`? `ts-esquery`? Etc.?

**Comment by auvred** [comment]
> roughly how many selectors was that 90%? I'm assuming it's a few hundred.

After a quick look through my changes: there is ~130 "old" selectors (just plain AST node types + AST node types with `:exit` at the end) and ~65 more complex selectors, from which I removed type annotations. 130 + 65 = ~195 types are currently dynamically inferred.

_In the group of "old" selectors I only included those that use `node` argument, because otherwise typescript doesn't calculate types for them. You can check this with `tsc --extendedDiagnostics` -> Instantiations_

So I didn't take into account the following listeners:

```ts
// ...
return {
  Identifier() { doSomeStuff() },
  MemberExpression: doSomeStuff
}
```

---

> Will it break anything if we were to just introduce it as a dependency now

In most cases, no. But there are a few places that can break, I think... For example `naming-convention`: https://github.com/typescript-eslint/typescript-eslint/blob/f36f68263c40a9baf7871eeece0f56f3d686edc5/packages/eslint-plugin/src/rules/naming-convention.ts#L686-L695

It returns object with type `{ [k: string]: (node: never) => void; }`. Unfortunately, it's a bit difficult to properly infer type of argument based on the property name:

```ts
createRule({
  create() {
    return {
      Identifier(node) { /* ... */ }
//      ^---------^-- we need to somehow link these two and consider that `Identifier` may be any string literal or string!     
    }
  }
})
```

Currently it works only if one of the listeners is a string literal.

Here is [TypeScript playground link](https://www.typescriptlang.org/play?moduleResolution=99&target=99&jsx=0&module=199&inlineSourceMap=true#code/JYWwDg9gTgLgBDAnmApnA3nAigVxVROAXzgDMoIQ4ByEAQwHNgBjAWhQGcBHPA6gKFCRYCZGkwAVAMoBRKRKgo0JcpRoABJKg7MowMDHYcANsAB2MAPRbOu-Yc4xFKanH78bcAPLGAJgCUcYxQAGWAOGBQzfA44AF4MfjhkuABJXyiYYFJgfAB+AC44AAozCAyi6TkFJQA6dMzs3KgASniAPjgANwhgXySUgFkUEAAjfBkADzBFDg5gCDNCkrKKuCr5Z1rhsYnp2fnFtrjOnr7+EgAyRJSaBosm-AKUSeAYamXS8pRK2U26+5ZHL4Y6nXr9W7UHbjKBTGacQ5mZ6vd6fVY-dZ-GoobYjGFwg4LMyg7rgi7ucyRKCkOjMNCBYJeAxE2LoAbJXQoOiRIoAHgkcBekTMvliET0ZgY7WKJLZt1uAG0ANJwczrAC6aO+RVw+EQvKVABpMdUtgA5b7tElnCEpIjkjLMYx0RRkHBmZhZRZwTnclAMlDFCDMxYcIoBpleswcFpFG3uSyWOAAQWMxlVsRy0X4vsiAeKcpSuYxMo6JUL8rSGQewKgXwybQrlZS6PZzeSiYAenk27ciIbeylqABhOhpgkIolwTq1Zhj4IoBXMShgHCRXxxGnGDgodXUesoRuDyut9u3Ls99v2vstC63-iJlNpjNkcwoHOKP35pvFoqlk43M2y7RvApgRFEMRFD4ARBKE4TCjE8SAWeVaNLWB6NnAlgAFRwLU+FwNhSbXs29rHnAigwDgUBmHAYEIVAHBtvaRD3o+MhQBQUAAITrGINCYAqADWRTiuYDDqn+6JFNEXQgmWNoANzEK44RwGU8B0HMwAMGYdCjMECAQKIqACXAyqqrREian+B46rw+pGnAFoZFaCngm0AA+cDuhkWYoL4ylENQtTFAATAAzGFYW3sW35tr+JSyuRwERHR8EQYxSHoCQWnIWewmiU44mSSs2rqSgcmtO55xXuRlHUbR9GZUxfZ3gmSYcVxvESPx1CCZMRUShJUnlXQZiINa4JBapsQaXAWnzLp+mGTAxmeP15kqmq1mfHZ2AOQaxouSgbkATaXk+SKKD+YFKmhZF0WxZ+eawQWCUvSWyXtqQ7qelOow4MAfhhOB0R1rGcBiZK+Xtg1NF3NWQLNAIdVwygVEI027YKoDwO+KDDEyuqGEYCRlbk8Qd5AA) with the reproduction of this issue. Any help will be appreciated :smile: 

The solution linked above looks very hacky, but I can't think of any other way to do this without changing the overall `createRule` signature :(

**Comment by auvred** [comment]
> Aside: `magic-esquery` is a very cute name (I find it very charming 😄) but not super descriptive. It'd be nice to have one that's more clear IMO. Especially since folks will inevitably want to dive into it to investigate type edge cases... `typed-esquery`? `esquery-types`? `ts-esquery`? Etc.?

I called it that, inspired by [`magic-regexp`](https://www.npmjs.com/package/magic-regexp) :sweat_smile:. But `typed-esquery` or `ts-esquery` or something similar sounds great too!  Let's consider renaming the package if the current name doesn't fit well!
